polkadot/runtime/parachains/src/assigner_coretime/mod.rs

```diff
@@ -387,67 +379,56 @@ impl<T: Config> Pallet<T> {
-	/// begin smaller than the one passed here. This restriction exists, because it makes the
-	/// insertion O(1) and the author could not think of a reason, why this restriction should be
-	/// causing any problems. Inserting arbitrarily causes a `DispatchError::DisallowedInsert`
-	/// error. This restriction could easily be lifted if need be and in fact an implementation is
-	/// available
-	/// [here](https://github.com/paritytech/polkadot-sdk/pull/1694/commits/c0c23b01fd2830910cde92c11960dad12cdff398#diff-0c85a46e448de79a5452395829986ee8747e17a857c27ab624304987d2dde8baR386).
-	/// The problem is that insertion complexity then depends on the size of the existing queue,
-	/// which makes determining weights hard and could lead to issues like overweight blocks (at
-	/// least in theory).
+	/// Important: Only appending is allowed or insertion into the last item. Meaning,
+	/// all already existing assignments must have a `begin` smaller or equal than the one passed
+	/// here.
+	/// Updating the last entry is supported to allow for making a core assignment multiple calls to
+	/// assign_core. Thus if you have too much interlacing for e.g. a single UMP message you can
+	/// split that up into multiple messages, each triggering a call to `assign_core`, together
+	/// forming the total assignment.
+	///
+	/// Inserting arbitrarily causes a `DispatchError::DisallowedInsert` error.
+	// With this restriction this function allows for O(1) complexity. It could easily be lifted, if
+	// need be and in fact an implementation is available
+	// [here](https://github.com/paritytech/polkadot-sdk/pull/1694/commits/c0c23b01fd2830910cde92c11960dad12cdff398#diff-0c85a46e448de79a5452395829986ee8747e17a857c27ab624304987d2dde8baR386).
+	// The problem is that insertion complexity then depends on the size of the existing queue,
+	// which makes determining weights hard and could lead to issues like overweight blocks (at
+	// least in theory).
 	pub fn assign_core(
 		core_idx: CoreIndex,
 		begin: BlockNumberFor<T>,
-		assignments: Vec<(CoreAssignment, PartsOf57600)>,
+		mut assignments: Vec<(CoreAssignment, PartsOf57600)>,
 		end_hint: Option<BlockNumberFor<T>>,
 	) -> Result<(), DispatchError> {
 		// There should be at least one assignment.
 		ensure!(!assignments.is_empty(), Error::<T>::AssignmentsEmpty);
 
-		// Checking for sort and unique manually, since we don't have access to iterator tools.
-		// This way of checking uniqueness only works since we also check sortedness.
-		assignments.iter().map(|x| &x.0).try_fold(None, |prev, cur| {
-			if prev.map_or(false, |p| p >= cur) {
-				Err(Error::<T>::AssignmentsNotSorted)
-			} else {
-				Ok(Some(cur))
-			}
-		})?;
-
-		// Check that the total parts between all assignments are equal to 57600
-		let parts_sum = assignments
-			.iter()
-			.map(|assignment| assignment.1)
-			.try_fold(PartsOf57600::ZERO, |sum, parts| {
-				sum.checked_add(parts).ok_or(Error::<T>::OverScheduled)
```

Comment:
isn't the OverScheduled error still worth checking for?

Reply:
Something to be checked by the Coretime chain. Erroring out here is not of much help. The scheduling logic does not really care either.